File: android/chesslibjava/src/com/chesslib/move/Move.java
package com.chesslib.move;

public abstract class Move {
    public static final int SQUARE_MASK = 63;
    public static final int PIECE_TYPE_MASK = 7;

    public static final int PROMOTE_TO_PIECE_CODE_SHIFT = 15;
    public static final int CAPTURED_PIECE_SHIFT = 18;
    public static final int CAPTURE_SHIFT = 21;

    public static final int CAPTURE = 1 << CAPTURE_SHIFT;
    public static final int EN_PASSANT_CAPTURE = 1 << 22;
    public static final int PROMOTION = 1 << 23;
    public static final int SHORT_CASTLE = 1 << 24;
    public static final int LONG_CASTLE = 1 << 25;
    public static final int PAWN_DOUBLE_MOVE = 1 << 26;

    public static final int ANY_CASTLE = SHORT_CASTLE | LONG_CASTLE;

    public static int getOriginSquare(int encodedMove) {
        return encodedMove & SQUARE_MASK;
    }

    public static int getDestSquare(int encodedMove) {
        return (encodedMove >> 6) & SQUARE_MASK;
    }

    public static int getPiece(int encodedMove) {
        return ((encodedMove >> 12) & PIECE_TYPE_MASK);
    }

    public static int getPromoteToPieceType(int encodedMove) {
        return ((encodedMove >> PROMOTE_TO_PIECE_CODE_SHIFT) & PIECE_TYPE_MASK);
    }

    public static int getCapturedPieceType(int encodedMove) {
        return ((encodedMove >> CAPTURED_PIECE_SHIFT) & PIECE_TYPE_MASK);
    }

    public static boolean isCapture(int encodedMove) {
        return (encodedMove & CAPTURE) != 0;
    }

    public static boolean isEnPassantCapture(int encodedMove) {
        return (encodedMove & EN_PASSANT_CAPTURE) != 0;
    }

    public static boolean isPromotion(int encodedMove) {
        return (encodedMove & PROMOTION) != 0;
    }

    public static boolean isShortCastle(int encodedMove) {
        return (encodedMove & SHORT_CASTLE) != 0;
    }

    public static boolean isLongCastle(int encodedMove) {
        return (encodedMove & LONG_CASTLE) != 0;
    }

    public static boolean isCastle(int encodedMove) {
        return (encodedMove & ANY_CASTLE) != 0;
    }

    public static boolean isPawnDoubleMove(int encodedMove) {
        return (encodedMove & PAWN_DOUBLE_MOVE) != 0;
    }
}

File: android/chesslibkt/src/main/java/ru/maksimbulva/chesslibkt/move/Move.kt
package ru.maksimbulva.chesslibkt.move

import ru.maksimbulva.chesslibkt.Piece
import ru.maksimbulva.chesslibkt.PieceEncoded
import ru.maksimbulva.chesslibkt.board.Square
import ru.maksimbulva.chesslibkt.pieceFromEncoded

val NullMove = Move(0)

// Use MoveBuilder class to get encoded value
inline class Move(
    /**
     * bits 0..5 - move origin square
     * bits 6..11 - move destination square
     * bits 12..14 - piece type to move
     * bits 15..17 - promote to piece type (pawn promotion)
     * bits 18..20 - captured piece type (for move undos)
     */
    private val encoded: Int
) {

    val originSquare: Square
        get() = Square(encoded and SQUARE_MASK)

    val destSquare: Square
        get() = Square((encoded shr 6) and SQUARE_MASK)

    private val pieceEncoded: Int
        get() = (encoded shr 12) and PIECE_TYPE_MASK

    val isPawnMove: Boolean
        get() = pieceEncoded == PieceEncoded.Pawn

//    piece_type_t getPieceType() const
//    {
//        return ;
//    }

//    square_t getCapturedPieceSquare() const
//    {
//        if (isEnPassantCapture()) {
//            // TODO: possibly this can be optimized
//            return encodeSquare(getRow(getOriginSquare()), getColumn(getDestSquare()));
//        }
//        else {
//            return getDestSquare();
//        }
//    }

    val promoteToPiece: Piece
        get() = when ((encoded shr 15) and PIECE_TYPE_MASK) {
            PieceEncoded.Knight -> Piece.Knight
            PieceEncoded.Bishop -> Piece.Bishop
            PieceEncoded.Rook -> Piece.Rook
            else -> Piece.Queen
        }

    val capturedPiece: Piece
        get() = when ((encoded shl 18) and PIECE_TYPE_MASK) {
            PieceEncoded.Pawn -> Piece.Pawn
            PieceEncoded.Knight -> Piece.Knight
            PieceEncoded.Bishop -> Piece.Bishop
            PieceEncoded.Rook -> Piece.Rook
            else -> Piece.Queen
        }

    val isCapture: Boolean
        get() = (encoded and Capture) != 0

    val isEnPassantCapture: Boolean
        get() = (encoded and EnPassantCapture) != 0

    val isPromotion: Boolean
        get() = (encoded and Promotion) != 0

    val isShortCastle: Boolean
        get() = (encoded and ShortCastle) != 0

    val isLongCastle: Boolean
        get() = (encoded and LongCastle) != 0

    val isCastle: Boolean
        get() = (encoded and AnyCastle) != 0

    val isPawnDoubleMove: Boolean
        get() = (encoded and PawnDoubleMove) != 0

    val isNullMove: Boolean
        get() = this == NullMove

    val getEncodedValue: Int
        get() = encoded

    companion object {
        const val Capture = 1 shl 21
        const val EnPassantCapture = 1 shl 22
        const val Promotion = 1 shl 23
        const val ShortCastle = 1 shl 24
        const val LongCastle = 1 shl 25
        const val PawnDoubleMove = 1 shl 26

        private const val AnyCastle = ShortCastle or LongCastle
    }
}

const val SQUARE_MASK = 63
const val PIECE_TYPE_MASK = 7
